coderd/agentapi/devcontainer_agent.go

```diff
@@ -0,0 +1,114 @@
+package agentapi
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+	"github.com/sqlc-dev/pqtype"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	agentproto "github.com/coder/coder/v2/agent/proto"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/provisioner"
+	"github.com/coder/quartz"
+)
+
+type DevContainerAgentAPI struct {
+	AgentID uuid.UUID
+	AgentFn func(context.Context) (database.WorkspaceAgent, error)
+
+	Log      slog.Logger
+	Clock    quartz.Clock
+	Database database.Store
+}
+
+func (a *DevContainerAgentAPI) CreateDevContainerAgent(ctx context.Context, req *agentproto.CreateDevContainerAgentRequest) (*agentproto.CreateDevContainerAgentResponse, error) {
+	//nolint:gocritic // We are the Dev Container Agent API, so this is safe.
+	ctx = dbauthz.AsDevContainerAgentAPI(ctx)
+
+	parentAgent, err := a.AgentFn(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("get parent agent: %w", err)
+	}
+
+	// NOTE(DanielleMaywood):
+	// This allows duplicate agent names to make it through.
+	// It would be nice if this was enforced at the database level.
+	agentName := req.Name
+	if agentName == "" {
+		return nil, xerrors.Errorf("agent name cannot be empty")
+	}
+	if !provisioner.AgentNameRegex.MatchString(agentName) {
+		return nil, xerrors.Errorf("agent name %q does not match regex %q", agentName, provisioner.AgentNameRegex.String())
+	}
+
+	devContainerAgent, err := a.Database.InsertWorkspaceAgent(ctx, database.InsertWorkspaceAgentParams{
+		ID:                       uuid.New(),
+		ParentID:                 uuid.NullUUID{Valid: true, UUID: parentAgent.ID},
+		CreatedAt:                a.Clock.Now(),
+		UpdatedAt:                a.Clock.Now(),
```

Comment:
Suggestion: assigning to a variable creates an atomic entry where created/updated is the exact same. Sometimes useful to know if it's been changed or not.